llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp

```diff
@@ -0,0 +1,96 @@
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace {
+struct FmulFaddMergePass : llvm::PassInfoMixin<FmulFaddMergePass> {
+  llvm::PreservedAnalyses run(llvm::Function &Func,
+                              llvm::FunctionAnalysisManager &) {
+    bool Changed = false;
+
+    struct ReplacementStruct {
+      llvm::BinaryOperator *FAdd;
+      llvm::BinaryOperator *FMul;
+      llvm::Value *ThirdOperand;
+    };
+
+    for (llvm::BasicBlock &BB : Func) {
+      std::vector<ReplacementStruct> ToReplace;
+
+      // 1st pass, collecting info about replacements
+      for (llvm::Instruction &Inst : BB) {
+        if (auto *FAdd = llvm::dyn_cast<llvm::BinaryOperator>(&Inst)) {
+          if (FAdd->getOpcode() == llvm::Instruction::FAdd) {
+            llvm::Value *Op0 = FAdd->getOperand(0);
+            llvm::Value *Op1 = FAdd->getOperand(1);
+
+            auto CheckOperand =
+                [&](llvm::Value *Operand,
```

Comment:
This lambda can be moved out of for-loops as invariant to iteration arguments.
```suggestion
            auto CheckOperand =
                [](llvm::Value *Operand,
```

Reply:
removed

Reply:
with moving lambda out of for-loops I'm getting an intrinsic in negative test, so what would we do? keep negative test or moving function instead?